File: src/model/Surface.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const errors = @import("../core/errors.zig");

pub const Parameter = struct {
    name: []const u8 = "",
    value: f64 = 0.0,

    pub fn validate(self: Parameter) errors.Error!void {
        if (self.name.len == 0 or !std.math.isFinite(self.value)) {
            return errors.Error.InvalidRequest;
        }
    }
};

pub const Surface = struct {
    pub const Kind = enum {
        lambertian,
        wavel_dependent,

        pub fn parse(value: []const u8) errors.Error!Kind {
            if (std.mem.eql(u8, value, "lambertian")) return .lambertian;
            if (std.mem.eql(u8, value, "wavel_dependent")) return .wavel_dependent;
            return errors.Error.InvalidRequest;
        }

        pub fn label(self: Kind) []const u8 {
            return @tagName(self);
        }
    };

    kind: Kind = .lambertian,
    // UNITS:
    //   `albedo` is a unitless hemispherical reflectance in the normalized `[0, 1]`
    //   range.
    albedo: f64 = 0.0,
    // UNITS:
    //   Optional surface pressure is expressed in hectopascals so canonical
    //   surface metadata can preserve the boundary pressure used by vendor-like
    //   interval and cloud-fraction configurations.
    pressure_hpa: f64 = 0.0,
    parameters: []const Parameter = &[_]Parameter{},

    pub fn validate(self: Surface) errors.Error!void {
        if (self.albedo < 0.0 or self.albedo > 1.0) {
            return errors.Error.InvalidRequest;
        }
        if (self.pressure_hpa != 0.0 and (!std.math.isFinite(self.pressure_hpa) or self.pressure_hpa <= 0.0)) {
            return errors.Error.InvalidRequest;
        }
        for (self.parameters) |parameter| {
            try parameter.validate();
        }
    }

    pub fn deinitOwned(self: *Surface, allocator: Allocator) void {
        if (self.parameters.len != 0) allocator.free(self.parameters);
        self.parameters = &[_]Parameter{};
    }
};

File: rust_src/input/surface.rs
use crate::common::errors;

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Parameter {
    pub name: String,
    pub value: f64,
}

impl Parameter {
    pub fn validate(&self) -> Result<(), errors::Error> {
        if self.name.is_empty() || !self.value.is_finite() {
            return Err(errors::Error::InvalidRequest);
        }
        Ok(())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    #[default]
    Lambertian,
    WavelDependent,
}

impl Kind {
    pub fn parse(value: &str) -> Result<Self, errors::Error> {
        match value {
            "lambertian" => Ok(Self::Lambertian),
            "wavel_dependent" => Ok(Self::WavelDependent),
            _ => Err(errors::Error::InvalidRequest),
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            Self::Lambertian => "lambertian",
            Self::WavelDependent => "wavel_dependent",
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Surface {
    pub kind: Kind,
    pub albedo: f64,
    pub pressure_hpa: f64,
    pub parameters: Vec<Parameter>,
}

impl Surface {
    pub fn validate(&self) -> Result<(), errors::Error> {
        // Albedo is unitless reflectance; validation only bounds finite values by compare.
        if self.albedo < 0.0 || self.albedo > 1.0 {
            return Err(errors::Error::InvalidRequest);
        }
        if self.pressure_hpa != 0.0 && (!self.pressure_hpa.is_finite() || self.pressure_hpa <= 0.0)
        {
            return Err(errors::Error::InvalidRequest);
        }
        for parameter in &self.parameters {
            parameter.validate()?;
        }
        Ok(())
    }
}
